### PR TITLE
8308076: X86_64: make rheapbase register allocatable in zero based compressedOops mode

### PR DESCRIPTION
**Pull request by kuaiwei**
In x86 64 mode, decode heap oop could use SIB without base if heap base is zero. like
```asm
0d1     movl    R11, [,R9 << 3 + #72] (zero base compressed oop addressing) # compressed ptr ! Field: java/lang/ClassLoader.classAssertionStatus 
```
So rheapbase( r12 )  can be allocated as general register.

Tier 1/2 tests are passed without new failure.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blocker
&nbsp;⚠️ Title mismatch between PR and JBS for issue [JDK-8308076](https://bugs.openjdk.org/browse/JDK-8308076)

### Issue
 * [JDK-8308076](https://bugs.openjdk.org/browse/JDK-8308076): make rheapbase register allocatable in zero based compressedOops mode for x86_64 (**Enhancement** - P4) ⚠️ Title mismatch between PR and JBS.

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13976/head:pull/13976` \
`$ git checkout pull/13976`

Update a local copy of the PR: \
`$ git checkout pull/13976` \
`$ git pull https://git.openjdk.org/jdk.git pull/13976/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13976`

View PR using the GUI difftool: \
`$ git pr show -t 13976`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13976.diff">https://git.openjdk.org/jdk/pull/13976.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13976#issuecomment-1547275191)

### CODE REVIEW

**Comment by merykitty** [comment]
Please see [JDK-8221249](https://bugs.openjdk.org/browse/JDK-8221249). A possible further improvement you can try is to allocate a vector register to be a dedicated zero register, this also helps other operations such as clear memory of newly created objects.

Thanks.

**Comment by kuaiwei** [comment]
> Please see [JDK-8221249](https://bugs.openjdk.org/browse/JDK-8221249). A possible further improvement you can try is to allocate a vector register to be a dedicated zero register, this also helps other operations such as clear memory of newly created objects.
> 
> Thanks.

Thanks for your suggestion. Before my work, I searched previous discussion but not find this one.

It looks no performance benefit in previous tests. But I find some performance gain in renaissance benchmark. And I also noticed some regression, in general, it looks a better score. I'm still running it multiple times to get a stable result. , I think the regression may has 2 causes
1 JVM can not use r12 as immediate 0
2 new heap decode instruction must has 32bit offset, it will enlarge the code size and increase icache pressure.

It's the tradeoff. It may dependents on app behavior. So I added a new option PreserveHeapbaseReg to switch.

File: src/hotspot/share/oops/compressedOops.hpp [hunk 1]
```diff
@@ -97,6 +98,12 @@ class CompressedOops : public AllStatic {
   static address  ptrs_base_addr()           { return (address)&_narrow_oop._base; }
   static address  ptrs_base()                { return _narrow_oop._base; }
 
+#if defined(X86) && !defined(ZERO)
```
Comment (merykitty) [inline src/hotspot/share/oops/compressedOops.hpp:101]: This should not be leaked into shared code

File: src/hotspot/cpu/x86/x86_64.ad [hunk 2]
```diff
@@ -635,68 +635,8 @@ void encode_RegMem(CodeBuffer &cbuf,
...
-        if (disp_reloc != relocInfo::none) {
-          emit_d32_reloc(cbuf, disp, relocInfo::oop_type, RELOC_DISP32);
-        } else {
-          emit_d32(cbuf, disp);
-        }
-      }
-    }
-  }
+  MacroAssembler masm(&cbuf);
+  masm.emit_regmem(reg, base, index, (Address::ScaleFactor)scale, disp, RelocationHolder::none);
```
Comment (merykitty) [inline src/hotspot/cpu/x86/x86_64.ad:639]: This change is not really related, right? I think a separate change to cleanup these would be more preferrable

**Review by merykitty — COMMENTED** [review]
Thanks for your explanations, my proposal can alleviate the usage of immediate 0 but will not solve the increased code size from base-less addresses. I have left some small comments.

Also, there is a `reinit_heapbase` in `MacroAssembler` that you may consider modifying as well.

File: src/hotspot/cpu/x86/x86_64.ad (hunk 2 above)
Comment (kuaiwei) [reply src/hotspot/cpu/x86/x86_64.ad:639]: The assembler function in adfile can not handle the case SIB without base reg. And emit_operand_helper in assembler_x86.cpp can cover more cases do more check ,like cpu type check in emit_compressed_disp_byte. So I think it's better to reuse the function in assembler_x86.cpp

File: src/hotspot/share/oops/compressedOops.hpp (hunk 1 above)
Comment (kuaiwei) [reply src/hotspot/share/oops/compressedOops.hpp]: Fixed.

**Comment by kuaiwei** [comment]
> 

Thanks for the comments. reinit_heapbase is fixed.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 3]
```diff
@@ -379,7 +379,7 @@ RegMask _STACK_OR_LONG_REG_mask;
 RegMask _STACK_OR_INT_REG_mask;
 
 static bool need_r12_heapbase() {
-  return UseCompressedOops;
+  return CompressedOops::need_heapbase_reg();
```
Comment (merykitty) [inline src/hotspot/cpu/x86/x86_64.ad:382]: Why not just use `PreserveHeapbaseReg`, this is a machine-dependent flag so I don't find it appealing to introduce a machine-independent check.

**Comment by theRealAph** [comment]
> It's the tradeoff. It may dependents on app behavior. So I added a new option PreserveHeapbaseReg to switch.

This looks like [JDK-8221249](https://bugs.openjdk.org/browse/JDK-8221249). Rather than starting a new patch, you should try to reopen that one.
